Add @CollectionTableOverride annotation for embeddable collections

RealTake opened the pull request:
### Add @CollectionTableOverride annotation for embeddable collections

This change was introduced to address a limitation in Hibernate where
entities cannot override the @CollectionTable(name = ...) defined inside an embeddable class.
When multiple entities share the same embeddable, it becomes impossible to customize the collection table name per entity, causing conflicts in DDL generation or preventing reuse of embeddable components.

To fix this, a new annotation (@CollectionTableOverride) was implemented, allowing entities to override the collection table name for collections declared inside embeddable classes.

- Add CollectionTableOverride and CollectionTableOverrides annotations
- Support overriding collection table name for collections within embeddable classes
- Implement field-level annotation lookup in CollectionBinder
- Improve path matching logic using BinderHelper.getRelativePath()
- Enable DDL generation and query generation with overridden table names

----------------------
By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt)
and can be relicensed under the terms of the [LGPL v2.1 license](https://www.gnu.org/licenses/old-licenses/lgpl-2.1.txt) in the future at the maintainers' discretion.
For more information on licensing, please check [here](https://github.com/hibernate/hibernate-orm/blob/main/CONTRIBUTING.md#legal).

----------------------

---
<!-- Hibernate GitHub Bot task list start -->
Please make sure that the following tasks are completed:
Tasks specific to HHH-19952 (New Feature):
- [x] Add tests for feature/improvement
- [ ] Update documentation as relevant: javadoc for changed API, `documentation/src/main/asciidoc/userguide` for all features, `documentation/src/main/asciidoc/introduction` for main features, links from existing documentation
- [x] Add entries as relevant to `migration-guide.adoc` (breaking changes) and `whats-new.adoc` (new features/improvements)


<!-- Hibernate GitHub Bot task list end -->

<!-- Hibernate GitHub Bot issue links start -->
<!-- THIS SECTION IS AUTOMATICALLY GENERATED, ANY MANUAL CHANGES WILL BE LOST -->
https://hibernate.atlassian.net/browse/HHH-19952
<!-- Hibernate GitHub Bot issue links end -->

RealTake committed:
HHH-19952: Add @CollectionTableOverride to override collection table names for embeddable collection fields

gavinking commented on hibernate-core/src/main/java/org/hibernate/boot/model/internal/CollectionBinder.java:
```diff
@@ -2805,4 +2819,149 @@ else if ( isManyToAny ) {
+
+		// Get the ClassDetails from the models context
+		final var modelsContext = buildingContext.getBootstrapContext().getModelsContext();
+		final ClassDetails entityClassDetails = modelsContext.getClassDetailsRegistry()
```
```suggestion
		final var entityClassDetails = modelsContext.getClassDetailsRegistry()
```

gavinking reviewed (changes requested):
Two requests:

1. we should allow the user to specify a whole `CollectionTable` nested annotation, and
2. the code formatting should be adjusted to follow the Hibernate conventions (spaces around parenthesized expressions).

RealTake committed:
HHH-19952: Apply code formatting

beikov commented on hibernate-core/src/test/java/org/hibernate/orm/test/bootstrap/binding/annotations/collectionTableOverride/CollectionTableOverrideTest.java:
```diff
@@ -0,0 +1,361 @@
+		private String name;
+
+		@Embedded
+		@CollectionTableOverride(name = "phones", table = "person_phones")
```
Did you try `@AssociationOverride(name = "phones", joinTable = @JoinTable(name = "person_phones"))`? It seems to me that the annotation would already serve the purpose for overriding the table name as well as join columns. I'd also say that using the existing annotation would be preferable to creating new ones.

beikov replied:
Well actually, I just found [this section of the spec](https://jakarta.ee/specifications/persistence/3.1/jakarta-persistence-spec-3.1#a14250):

>To override the default properties of the column used for a basic type, the Column annotation is used on the collection-valued attribute in addition to the ElementCollection annotation. The value of the table element of the Column annotation defaults to the name of the collection table.

>To override these defaults for an embeddable class, the AttributeOverride and/or AttributeOverrides annotations must be used in addition to the ElementCollection annotation. The value of the table element of the Column annotation used in the AttributeOverride annotation defaults to the name of the collection table. If the embeddable class contains references to other entities, the default values for the columns corresponding to those references may be overridden by means of the AssociationOverride and/or AssociationOverrides annotations.

So I guess by overriding every column with `@AttributeOverride` and specifying a table it would be overridden?

beikov replied:
But then there is also [this section](https://jakarta.ee/specifications/persistence/3.1/jakarta-persistence-spec-3.1#collectiontable):

>A collection table is created for the mapping of an element collection. See [Section 11.1.8](https://jakarta.ee/specifications/persistence/3.1/jakarta-persistence-spec-3.1#a14250) for the rules that apply to the generation of collection tables. The Column, AttributeOverride, and AssociationOverride annotations may be used to override CollectionTable mappings, as described in [Section 11.1.9](https://jakarta.ee/specifications/persistence/3.1/jakarta-persistence-spec-3.1#a14330), [Section 11.1.4](https://jakarta.ee/specifications/persistence/3.1/jakarta-persistence-spec-3.1#a14084), and [Section 11.1.2](https://jakarta.ee/specifications/persistence/3.1/jakarta-persistence-spec-3.1#a13942) respectively.

gavinking replied:
Pfffffffffff shit I have little idea what that means.

But I don't *think* it means what you're suggestion. I think it's much more likely talking about the case where you have the embeddable as the *element* of the collection.

gavinking commented on hibernate-core/src/main/java/org/hibernate/boot/model/internal/CollectionBinder.java:
```diff
@@ -2805,4 +2819,149 @@ else if ( isManyToAny ) {
+		final String fieldName;
+		if (fullPath.contains(".")) {
+			fieldName = fullPath.substring(0, fullPath.indexOf('.'));
+		} else {
```
```suggestion
		}
		else {
```

RealTake committed:
HHH-19952: Support nested @CollectionTable in @CollectionTableOverride

Replace table name string with nested @CollectionTable annotation
to allow full collection table configuration override, including
schema, catalog, joinColumns, indexes, and other attributes.

RealTake committed:
HHH-19952: Update hibernate-core/src/main/java/org/hibernate/boot/model/internal/CollectionBinder.java

Co-authored-by: Gavin King <gavin@hibernate.org>

RealTake committed:
HHH-19952: Use var for member iteration in CollectionBinder

RealTake committed:
HHH-19952: Dynamic Type error fixed.

Commit:
HHH-19952: Merge hibernate/main into RealTake/main

RealTake committed:
Merge branch 'main' into main
